include/swift/Parse/ASTGen.h

```diff
@@ -14,54 +14,107 @@
 #define SWIFT_PARSE_ASTGEN_H
 
 #include "swift/AST/ASTContext.h"
+#include "swift/AST/Decl.h"
 #include "swift/AST/Expr.h"
+#include "swift/Parse/PersistentParserState.h"
 #include "swift/Syntax/SyntaxNodes.h"
 #include "llvm/ADT/DenseMap.h"
 
 namespace swift {
 /// Generates AST nodes from Syntax nodes.
 class ASTGen {
   ASTContext &Context;
-  // A stack of source locations of syntax constructs. Allows us to get the
-  // SourceLoc necessary to create AST nodes for nodes in not-yet-complete
-  // Syntax tree. The topmost item should always correspond to the token/node
-  // that has been parsed/transformed most recently.
-  // todo [gsoc]: remove when possible
-  llvm::SmallVector<SourceLoc, 16> LocStack;
+
+  /// Type cache to prevent multiple transformations of the same syntax node.
+  llvm::DenseMap<syntax::SyntaxNodeId, TypeRepr *> TypeCache;
+
+  PersistentParserState **ParserState;
+
+  /// Types that are not representable as syntax nodes.
+  llvm::DenseMap<const void *, TypeRepr *> Types;
```

Comment:
If the key is `SourceLoc` we should be able to use it as the key. i.e. llvm::DenseMap<SourceLoc, TypeRepr *>

Reply:
I think this is to hold `SILBoxTypeRepr` (for now). Could you add `FIXME` doc-comment to we should remove this eventually?

Reply:
I didn't realise `SourceLoc` is trivially hashable (it only contains a single hashable member).
I've updated the key type, but for the value type, we have to keep `TypeRepr *`.
Currently, `SILBoxTypeRepr` cannot be represented by Syntax *and* `ASTGen` cannot generate all attributed types.
That's because `parseTypeAttributeList` method is part of declaration parsing I didn't refactor yet and therefore we don't have the corresponding code in `ASTGen` either.
I've updated the comment to reflex this and added the fixme.